### PR TITLE
Change ArraySortHelper to use Comparison<T>

### LINKED ISSUE #8481: ArraySortHelper is not exactly efficient when Comparison<T> is used
In my experience sorting an array (or list) of objects using something like `Sort((x, y) => x.Age.CompareTo(y.Age)` is quite common these days. Building a custom comparer object is usually a waste of time and as we'll see it is also less efficient than using `Comparison<T>`.

But `ArraySortHelper` is based on `IComparer<T>` rather than `Comparison<T>`. So:
* one way or another you'll pay the cost of an interface call for each compare operation
* if you're using a comparison delegate you'll path both cost of the interface call and the delegate call as the delegate gets wrapped in a comparer object

Here's a quick experiment:
```C#
class Person
{
    public int Age { get; set; }
}

class PersonComparer : IComparer<Person>
{
    public int Compare(Person x, Person y) => x.Age.CompareTo(y.Age);
}

static void Main()
{
    var rnd = new Random(42);
    Person[] array = (from x in Enumerable.Range(1, 10000) select new Person { Age = rnd.Next() }).ToArray();
    IComparer<Person> pc = new PersonComparer();
    IComparer<Person> dc = Comparer<Person>.Create((x, y) => x.Age.CompareTo(y.Age));
    Comparison<Person> dpc = pc.Compare;
    Comparison<Person> ddc = (x, y) => x.Age.CompareTo(y.Age);

    var sw = Stopwatch.StartNew();

    //ArraySortHelper<Person>.Default.Sort(array, 0, array.Length, pc);
    //ArraySortHelper<Person>.Default.Sort(array, 0, array.Length, dc);
    //ArraySortHelperDelegate<Person>.Default.Sort(array, 0, array.Length, dpc);
    //ArraySortHelperDelegate<Person>.Default.Sort(array, 0, array.Length, ddc);

    Console.WriteLine(sw.ElapsedTicks);
}
```
Version | ~Ticks | Compare overhead
---|---|---
ArraySortHelper + Comparer | 4800 | 1 interface call
ArraySortHelper + Comparison | 5200 | 1 interface call + 1 delegate call
ArraySortHelperDelegate + Comparer | 3600 | 1 delegate call
ArraySortHelperDelegate + Comparison | 3600 | 1 delegate call

`ArrayShortHelperDelegate` is a version of the original `ArraySortHelper` that uses `Comparison<T>` instead of `IComparer<T>`.

Notice that when `ArrayShortHelperDelegate` is used it makes no difference if a comparer or a comparison is used. That's because the cost of interface lookup is paid only once, at delegate creation time.

Now, it's easy to add `ArraySortHelperDelegate` and route the existing `Sort(Comparison<T>)` overloads to it but I think it would be better to just change the existing `ArraySortHelper`. If that's not possible would the resulting code duplication be acceptable?

Are there any other concerns about such a change? Sorting has a history of compatibility issue but I don't think replacing `IComparer<T>` with `Comparison<T>` will cause any issues, it's the same algorithm and it will react the same way to poorly implemented comparers/comparisons.

### PR DESCRIPTION
**Pull request by mikedn**
The Array/List.Sort overloads that take a `Comparison<T>` have worse performance than the ones that take a `IComparer<T>`. That's because sorting is implemented around `IComparer<T>` and a `Comparison<T>` needs to be wrapped in a comparer object to be used.

At the same time, interface calls are slower than delegate calls so the existing implementation doesn't offer the best performance even when the `IComparer<T>` based overloads are used.

By changing the implementation to use `Comparison<T>` we avoid interface calls in both cases.

When `IComparer<T>` overloads are used a `Comparison<T>` delegate is created from `IComparer<T>.Compare`, that's an extra object allocation but sorting is faster and we avoid having two separate sorting implementations.

Fixes #8481

### CODE REVIEW

File: src/mscorlib/src/System/Collections/Generic/ArraySortHelper.cs [hunk 1]
```diff
@@ -148,6 +148,42 @@ public int BinarySearch(T[] array, int index, int length, T value, IComparer<T>
...
+            try
+            {
+#if FEATURE_CORECLR
+                // Since QuickSort and IntrospectiveSort produce different sorting sequence for equal keys the upgrade 
+                // to IntrospectiveSort was quirked. However since the phone builds always shipped with the new sort aka 
+                // IntrospectiveSort and we would want to continue using this sort moving forward CoreCLR always uses the new sort.
+
+                IntrospectiveSort(keys, index, length, comparer);
+#else
+                if (BinaryCompatibility.TargetsAtLeast_Desktop_V4_5)
```
Comment (jkotas) [inline src/mscorlib/src/System/Collections/Generic/ArraySortHelper.cs]: Yes, it is fine to delete the code under `#if !FEATURE_CORECLR`. https://github.com/dotnet/coreclr/issues/972#issuecomment-257633318

**Comment by jkotas** [comment]
Looks pretty good to me.

**Comment by mikedn** [comment]
I suspect that there's more that can be done about the default comparer case, e.g. caching the delegate built from the default comparer. I doubt it's worth it and can be done in a separate PR anyway.

I didn't modify `ArraySortHelper<T>.BinarySearch` because `BinarySearch` doesn't currently offer a `Comparison<T>` overload. Add allocating a delegate every time you search using a comparer doesn't sound like good compromise.

I also didn't modify `ArraySortHelper<TKey, TValue>`. The key/value `Sort` overloads also don't offer `Comparison<T>` versions and in my experience they are rarely used.

**Comment by jkotas** [comment]
Thanks!

**Comment by jkotas** [comment]
Would you mind porting the change to CoreRT as well?

**Pull request merged**